doc/design/block_execution.md

```diff
@@ -0,0 +1,113 @@
+
+1. Baseline
+  - a single thread runs a block on a single device, e.g., a CPU or a GPU.
+
+2. Data Parallelism
+  - multi-threads in a stand-alone machine: Only one copy of parameters needs to be storaged in CPU memory.
+  - multi-GPUs in a stand-alone machine: Each GPU card will have a full copy of parameters in its own GPU memory.
+  - multi-threads/multi-GPUs in a cluster: Each node in the cluster supports multi-threads/multi-GPUs.
+
+3. Model Parallelism
+  - Operators in a block can locate in different devices(CPU/GPU/FPGA) in a stand-alone machine. Users have to set device id for every operator manually.
+  - We should notice another situation where operators in a block can be executed in different CPU threads or CUDA streams to achieve high performace, but they are actually in the same device. It's scheduled by Paddle automatically, and users don't need to set specific CPU thread id or CUDA stream id. 
+
+4. Long-term
+  - Paddle schedules CPU/GPU/FPGA in a stand-alone machine automatically, makes full use of them to get best performance when executing a neural network.
+  - Paddle schedules CPU/GPU/FPGA in a cluster automatically, and overlaps communication and computation well to get best performance.
+
+  
+  
+## Analysis 
+
+
+Since a block is actually a graph, the data member of a graph should be Nodes and Edges. Every Node must have a device id for descirbing the place information. Please refer to the design of [TensorFlow](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/node_def.proto#L48). In our block design now, the data members are VarDesc and OpDesc. So, VarDesc and OpDesc must have a field to descirbe concrete device information. 
+
+Let's consider about this question, What configuration should Paddle get from uses to execute a block in various hardward environment.
+
+  
+- Data Parallelism
+  - stand-alone machine: Users have to set several device ids, and each device will execute the same block.
+  - cluster: Users have to set node ids and device ids, and each device will execute the same block.
+  
+- Model Parallelism
+  - Users have to set a concrete device id for each operator in a topology. Paddle will first execute the operator strictly following users' configuration, and will do some another parallel optimization automatically.
+
+ 
+
+
+## Solution
+  
+Actually, we can have both data parallelism and model parallelism. In order to support the mixing paralelism, we have to propose these two strategies:
+
+
+- A simple device id setting rules
+  - Firsly, users have to set one or several default device id for a topology. These devices are the same and usually finish the most computation of a topology. We mainly consider data parallelism in this step.
+  - Secondly, users need to set device id for each operators in a topology. If users set nothing, the operator will just use default device id. When devices have to be switched, users have to add a copy operator manually.
```

Comment:
> users have to add a copy operator manually

I think the copy operator should be transparent to the user.

We need a interface that allow user to specify **what** they want, not **how** they want things to be done. Only in this way, the framework will have the freedom to do optimizations.

Reply:
Yes, I think that we should allow users change variable's device and copy operator is needed.